src/tree/updater_gpu_hist.cu

```diff
@@ -402,34 +402,34 @@ struct CalcWeightTrainParam {
 
 template <typename GradientSumT>
 __global__ void SharedMemHistKernel(xgboost::EllpackMatrix matrix,
+                                    size_t base_rowid,
                                     common::Span<const RowPartitioner::RowIndexT> d_ridx,
                                     GradientSumT* d_node_hist,
                                     const GradientPair* d_gpair, size_t n_elements,
                                     bool use_shared_memory_histograms) {
   extern __shared__ char smem[];
   GradientSumT* smem_arr = reinterpret_cast<GradientSumT*>(smem);  // NOLINT
   if (use_shared_memory_histograms) {
-    dh::BlockFill(smem_arr, matrix.BinCount(), GradientSumT());
+    dh::BlockFill(smem_arr, matrix.info.n_bins, GradientSumT());
     __syncthreads();
   }
   for (auto idx : dh::GridStrideRange(static_cast<size_t>(0), n_elements)) {
-    int ridx = d_ridx[idx / matrix.info.row_stride ];
+    int ridx = d_ridx[idx / matrix.info.row_stride];
     int gidx =
         matrix.gidx_iter[ridx * matrix.info.row_stride + idx % matrix.info.row_stride];
     if (gidx != matrix.info.n_bins) {
       // If we are not using shared memory, accumulate the values directly into
       // global memory
       GradientSumT* atomic_add_ptr =
           use_shared_memory_histograms ? smem_arr : d_node_hist;
-      dh::AtomicAddGpair(atomic_add_ptr + gidx, d_gpair[ridx]);
+      dh::AtomicAddGpair(atomic_add_ptr + gidx, d_gpair[ridx + base_rowid]);
```

Comment:
I think this can be tidied up a little. `base_rowid` can be a member of EllpackMatrix. EllpackMatrix could have a method that returns the row index of each of its elements.

Reply:
Done.